src/backend/columnar/columnar_customscan.c

```diff
@@ -228,6 +244,172 @@ IsNotIndexPath(Path *path)
+	indexPath->path.total_cost = indexPath->path.startup_cost +
+								 indexAMScanCost + columnarIndexScanCost;
+
+	ereport(DEBUG4, (errmsg("columnar table index scan costs re-estimated "
+							"by columnarAM (including indexAM costs): "
+							"startup cost = %.10f, total cost = %.10f",
+							indexPath->path.startup_cost,
+							indexPath->path.total_cost)));
+}
+
+
+/*
+ * ColumnarIndexScanAddStartupCost returns additional startup cost estimated
+ * for index scan described by IndexPath for columnar table with relationId.
+ */
+static Cost
+ColumnarIndexScanAddStartupCost(RelOptInfo *rel, Oid relationId, IndexPath *indexPath)
+{
+	int numberOfColumnsRead = RelationIdGetNumberOfAttributes(relationId);
+
+	/* we would at least read one stripe */
+	return ColumnarPerStripeScanCost(rel, relationId, numberOfColumnsRead);
+}
+
+
+/*
+ * ColumnarIndexScanAddTotalCost returns additional cost estimated for
+ * index scan described by IndexPath for columnar table with relationId.
+ */
+static Cost
+ColumnarIndexScanAddTotalCost(PlannerInfo *root, RelOptInfo *rel,
+							  Oid relationId, IndexPath *indexPath)
+{
+	int numberOfColumnsRead = RelationIdGetNumberOfAttributes(relationId);
+	Cost perStripeCost = ColumnarPerStripeScanCost(rel, relationId, numberOfColumnsRead);
+
+	/*
+	 * We don't need to pass correct loop count to amcostestimate since we
+	 * will only use index correlation & index selectivity, and loop count
+	 * doesn't have any effect on those two.
+	 */
+	double fakeLoopCount = 1;
+	Cost fakeIndexStartupCost;
+	Cost fakeIndexTotalCost;
+	double fakeIndexPages;
+	Selectivity indexSelectivity;
+	double indexCorrelation;
+	amcostestimate_function amcostestimate = indexPath->indexinfo->amcostestimate;
+	amcostestimate(root, indexPath, fakeLoopCount, &fakeIndexStartupCost,
+				   &fakeIndexTotalCost, &indexSelectivity,
+				   &indexCorrelation, &fakeIndexPages);
+
+	Relation relation = RelationIdGetRelation(relationId);
+	uint64 rowCount = ColumnarTableRowCount(relation);
+	RelationClose(relation);
+	double estimatedRows = rowCount * indexSelectivity;
+
+	/*
+	 * In the worst case (i.e no correlation between the column & the index),
+	 * we need to read a different stripe for each row.
+	 */
+	double maxStripeReadCount = estimatedRows;
+
+	/*
+	 * In the best case (i.e the column is fully correlated with the index),
+	 * we wouldn't read the same stripe again and again thanks
+	 * to locality.
+	 */
+	double avgStripeRowCount =
+		rowCount / (double) ColumnarTableStripeCount(relationId);
+	double minStripeReadCount = estimatedRows / avgStripeRowCount;
+
+	/*
+	 * While being close to 0 means low correlation, being close to -1 or +1
+	 * means high correlation. For index scans on columnar tables, it doesn't
+	 * matter if the column and the index are "correlated" (+1) or
+	 * "anti-correlated" (-1) since both help us avoiding from reading the
+	 * same chunk group again and again.
+	 */
```

Comment:
Should we be counting the number of stripe reads, or chunk reads, or both? I'm inclined to think chunk reads are more important because of the decompression, but stripe reads do have some overhead as well due to missing blocks in the cache and not benefiting from readahead.

Reply:
Initially I was thinking of calculating the cost based on number of chunk groups being decompressed.

However, reading the whole stripe (regardless of how many chunk groups we need to decompress) for a single tuple seemed to be the heavy weight part to me.

Also realized that, for custom scan, we were only taking number of blocks read from disk, not the decompressed chunk groups.

For those reasons, I followed a similar approach for index scan too.

On the other hand, ideally, we should also consider chunk group decompression cost for both custom scan & index scan.
But I think we can ignore that for the time being.

What do you think @jeff-davis ?